woocommerce/woocommerce-ios
# [Hack Week][Widgets][Do not merge]: Implement Home Screen Widgets

This Pull Request is not intended to be merged: It is too big and has few hacks that are not acceptable on trunk. Instead I want to use it to show the code and gather feedback and suggestions. That way we can create a project and start working on a real implementation with smaller issues/Pull Requests.



### Description

This Pull Request implements the second iteration of the Figma design for the simple Stats Home Screen widget (See screenshots). To achieve that I:

- Created a new Widget Extension
- Implemented sharing data between the main app and the Widgets Extension
- Fetched new data from the extension using the Networking remotes
- Implement the design for Today and This Week widgets, in small and medium size.
- Added a prototype of Feature flagging (see comments).
- Implemented the widgets placeholder preview. 

Please take a look here for more information.

### Testing instructions

Debugging this on the simulator is very unstable at the moment, so it is truly encouraged to do so on device for a better experience. Please notice that, since we didn't have a dedicated Woo Provisioning profile for this extension I had to use the ones from WordPress, updating the entitlements so it might work.

Prerequisites: User is logged in and a store selected.

1. On Xcode, select the WooCommerceStatsWidgetExtension
2. Run it. Again, Xcode is very flawed when working with the Widgets extension. It was my experience that I had to remove the Woo Widgets from the Homescreen and Today view if I wanted the Xcode debugger to be attached to this process.
3. Home screen Today-small widget is shown. 
4. If you want to add more widgets types:
- In the Home Screen, scroll left to the Today view.
- Tap on Edit
- Tap on + 
- Scroll down to Woo
- Select the desired widget

### Screenshots

<img src="https://user-images.githubusercontent.com/1864060/174341919-dd93e800-8804-4fbd-9b0b-acd3a337529e.PNG" width="375">
<img src="https://user-images.githubusercontent.com/1864060/174342022-e1cc5799-54bb-4629-bcfa-5c91e1e19fef.PNG" width="375">
<img src="https://user-images.githubusercontent.com/1864060/174342100-804c8422-d90e-4574-8df2-afb2b3409a91.PNG" width="375">

---
